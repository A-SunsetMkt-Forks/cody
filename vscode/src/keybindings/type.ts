import type { CommentArray } from 'comment-json'

export type KeybindingsContent = CommentArray<Record<string, string>> | null | undefined
