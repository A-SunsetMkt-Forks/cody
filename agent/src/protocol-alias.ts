export * from '../../vscode/src/jsonrpc/agent-protocol'
export * as protocolFactory from '../../vscode/src/jsonrpc/agent-protocol-factory'
