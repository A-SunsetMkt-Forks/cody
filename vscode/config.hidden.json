{
  "type": "object",
  "title": "Cody Hidden",
  "properties": {}
}
