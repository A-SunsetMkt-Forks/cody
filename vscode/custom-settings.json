{
    "cody.autocomplete.advanced.provider": "unstable-codegen",
    "cody.autocomplete.advanced.serverEndpoint": "https://backend4.codegen.salesforceresearch.ai/complete_batch",
    "cody.autocomplete.enabled": true,
    "cody.autocomplete.advanced.embeddings": true,

    "cody.useContext": "embeddings",
    "cody.experimental.chatPredictions": false,
    "cody.experimental.guardrails": false,
    "cody.plugins.enabled": false,
    "cody.inlineChat.enabled": false
}
