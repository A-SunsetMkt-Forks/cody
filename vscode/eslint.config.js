// @ts-check

const tslint = require('typescript-eslint')

module.exports = {
    languageOptions: {
        parser: tslint.parser,
        parserOptions: {
            project: ['tsconfig.json'],
        },
    },
    files: ['src/**/*.ts', 'src/**/*.tsx'],
    ignores: ['**/tree-sitter/**'],
    plugins: {
        '@typescript-eslint': tslint.plugin,
    },
    rules: {
        '@typescript-eslint/strict-boolean-expressions': ['error', {
            allowString: false,
            allowNumber: true,
            allowNullableObject: true,
            allowNullableBoolean: true,
            allowNullableString: true,
            allowNullableNumber: true,
            allowNullableEnum: true,
            allowAny: true,
            allowRuleToRunWithoutStrictNullChecksIKnowWhatIAmDoing: true,
        }],
    },
}
