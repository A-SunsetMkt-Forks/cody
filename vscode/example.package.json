"openctx.providers": {
          "type": "object",
          "markdownDescription": "OpenCtx providers configuration.",
          "default": {},
          "scope": "resource",
          "properties": {
            "A": {
              "type": "string"
            },
            "B": {
              "type": "string"
            }
          }
        },
