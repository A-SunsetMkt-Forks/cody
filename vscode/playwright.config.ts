import { defineConfig } from '@playwright/test'

export default defineConfig({
    workers: 1,
    // Give failing tests a second chance
    retries: 0, // TODO(sqs): back to 2
    testDir: 'test/e2e',
    expect: {
        timeout: 2000,
    },
})
