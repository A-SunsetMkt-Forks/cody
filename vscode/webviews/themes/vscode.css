body.vscode-dark {
    background: transparent; /* in vscode we should let the color of the underlying UI set the background of the sidebar*/
    /* Override VS Code Webview Toolkit elements */
    --border-width: none;
    color:var(--text);

    --font: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;
    --font-editor: 'Monaco', 'Consolas', monospace;
    --font-size: 13px;

    --vscode-font-family: var(--font);

    /* PRIMITIVES FROM EDITOR */
    --editor-background: var(--vscode-panel-background);
    --editor-border: var(--vscode-panel-border);
    --editor-accent: var(--vscode-button-background);
    --editor-text:var(--vscode-foreground);

    /* Primary Button */
    --button-primary-text: var(--vscode-button-foreground);
    --button-primary-background: var(--editor-accent);
    --button-primary-background-hover: var(--vscode-button-hoverBackground);
    --button-primary-border: var(--vscode-button-border);
    --button-primary-border-hover: var(--vscode-button-border);

    /* Backgrounds */
    --background-01: var(--editor-background);  /* Main page background */
    --background-02: color-mix(in srgb, var(--background-01), black 5%);  /* Slightly darker */
    --background-03: color-mix(in srgb, var(--background-01), black 10%);  /* Even darker */

    /* Text */
    --text: var(--editor-text);
    --text-subtle: color-mix(in srgb, var(--text) 70%, transparent 30%);  /* Slightly transparent */
    --text-muted: color-mix(in srgb, var(--text) 50%, transparent 50%);  /* More transparent */
    --text-primary: var(--editor-text);
    --text-primary-subtle: color-mix(in srgb, var(--text-primary)  70%, transparent 30%);  /* Subtle primary */
    --text-primary-muted: color-mix(in srgb, var(--text-primary)  50%, transparent 50%);  /* Muted primary */
    --text-disabled: color-mix(in srgb, var(--text-primary) 30%, transparent 70%);  /* Highly transparent */

    /* Links */
    --text-link-color: var(--editor-accent);
    --text-link-hover-color: #ff4c00;
    --text-white: #FFFFFF;

    /* Borders */
    --border: var(--editor-border);
    --border-subtle: color-mix(in srgb, var(--border) 70%, transparent 30%);  /* Subtle border */
    --border-disabled: color-mix(in srgb, var(--border) 30%, transparent 70%);  /* Disabled border */
    --border-active: color-mix(in srgb, var(--editor-accent), transparent 30%);  /* Active border with hint of blue */
    --border-hover: color-mix(in srgb, var(--border) 75%, white 25%);  /* Hover border with more vibrant blue */

    /* Highlight */
    --highlight-background: color-mix(in srgb, var(--editor-accent), transparent 40%);
    --highlight-text: color-mix(in srgb, var(--text), white 90%);
    --highlight-border: color-mix(in srgb, var(--editor-accent) 50%, white 20%);

    /* Links */
    --link-color: var(--editor-accent);
    --link-hover-color: color-mix(in srgb, var(--editor-accent) 75%, white 25%);
}

body.vscode-light {
    background:pink !important;
}
