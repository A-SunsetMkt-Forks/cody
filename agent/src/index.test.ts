import assert from 'assert'
import { execSync, spawn } from 'child_process'
import fspromises from 'fs/promises'
import path from 'path'

import * as uuid from 'uuid'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import * as vscode from 'vscode'
import { Uri } from 'vscode'

import type { ExtensionMessage } from '../../vscode/src/chat/protocol'

import { MessageHandler } from './jsonrpc-alias'
import { ClientInfo, ServerInfo } from './protocol-alias'

export class TestClient extends MessageHandler {
    constructor() {
        super()
        this.registerNotification('debug/message', message => {
            console.log(`${message.channel}: ${message.message}`)
        })
    }

    public async handshake(clientInfo: ClientInfo): Promise<ServerInfo> {
        return new Promise((resolve, reject) => {
            setTimeout(reject, 1000)
            this.request('initialize', clientInfo).then(
                info => {
                    this.notify('initialized', null)
                    resolve(info)
                },
                error => reject(error)
            )
        })
    }

    public async shutdownAndExit() {
        await this.request('shutdown', null)
        this.notify('exit', null)
    }
}

const workspaceRootUri = vscode.Uri.parse('file:///Users/olafurpg/dev/sourcegraph/bfg-demo')
const workspaceRootPath = workspaceRootUri.fsPath
const dotcom = 'https://sourcegraph.com'
const clientInfo: ClientInfo = {
    name: 'test-client',
    version: 'v1',
    workspaceRootUri: workspaceRootUri.toString(),
    workspaceRootPath,
    extensionConfiguration: {
        anonymousUserID: 'abcde1234',
        accessToken: process.env.SRC_ACCESS_TOKEN ?? 'sgp_RRRRRRRREEEEEEEDDDDDDAAACCCCCTEEEEEEEDDD',
        serverEndpoint: dotcom,
        customHeaders: {},
        autocompleteAdvancedProvider: 'anthropic',
        autocompleteAdvancedAccessToken: '',
        autocompleteAdvancedServerEndpoint: '',
        debug: false,
        verboseDebug: false,
        codebase: 'github.com/sourcegraph/cody',
    },
}

const cwd = process.cwd()
const agentDir = path.basename(cwd) === 'agent' ? cwd : path.join(cwd, 'agent')
const recordingDirectory = path.join(agentDir, 'recordings')
const agentScript = path.join(agentDir, 'dist', 'index.js')

if (process.env.CODY_RECORDING_MODE === 'record' || process.env.CODY_RECORD_IF_MISSING === 'true') {
    console.log('Because CODY_RECORDING_MODE=record, validating that you are authenticated to sourcegraph.com')
    execSync('src login', { stdio: 'inherit' })
    assert.strictEqual(
        process.env.SRC_ENDPOINT,
        clientInfo.extensionConfiguration?.serverEndpoint,
        'SRC_ENDPOINT must match clientInfo.extensionConfiguration.serverEndpoint'
    )
}

describe('Agent', () => {
    // Uncomment the code block below to disable agent tests. Feel free to do this to unblock
    // merging a PR if the agent tests are failing. If you decide to uncomment this block, please
    // post in #wg-cody-agent to let the team know the tests have been disabled so that we can
    // investigate the problem and get the passing again.
    // if (process.env.SRC_ACCESS_TOKEN === undefined || process.env.SRC_ENDPOINT === undefined) {
    //     it('no-op test because SRC_ACCESS_TOKEN is not set. To actually run the Cody Agent tests, set the environment variables SRC_ENDPOINT and SRC_ACCESS_TOKEN', () => {})
    //     return
    // }

    if (process.env.VITEST_ONLY && !process.env.VITEST_ONLY.includes('Agent')) {
        it('Agent tests are skipped due to VITEST_ONLY environment variable', () => {})
        return
    }
    const client = new TestClient()

    // Bundle the agent. When running `pnpm run test`, vitest doesn't re-run this step.
    execSync('pnpm run build', { cwd: agentDir, stdio: 'inherit' })

    const agentProcess = spawn('node', ['--enable-source-maps', '--inspect', agentScript, 'jsonrpc'], {
        stdio: 'pipe',
        cwd: agentDir,
        env: {
            CODY_SHIM_TESTING: 'true',
            CODY_RECORDING_MODE: 'replay', // can be overwritten with process.env.CODY_RECORDING_MODE
            CODY_RECORDING_DIRECTORY: recordingDirectory,
            CODY_RECORDING_NAME: 'FullConfig',
            ...process.env,
        },
    })
    client.connectProcess(agentProcess, error => {
        console.log({ error })
        process.exit(1)
    })

    // Initialize inside beforeAll so that subsequent tests are skipped if initialization fails.
    beforeAll(async () => {
        const serverInfo = await client.handshake(clientInfo)
        assert.deepStrictEqual(serverInfo.name, 'cody-agent', 'Agent should be cody-agent')
    }, 5000)

    it('handles config changes correctly', () => {
        // Send two config change notifications because this is what the
        // JetBrains client does and there was a bug where everything worked
        // fine as long as we didn't send the second unauthenticated config
        // change.
        client.notify('extensionConfiguration/didChange', {
            ...clientInfo.extensionConfiguration,
            anonymousUserID: 'abcde1234',
            accessToken: '',
            serverEndpoint: 'https://sourcegraph.com/',
            customHeaders: {},
        })
        client.notify('extensionConfiguration/didChange', {
            ...clientInfo.extensionConfiguration,
            anonymousUserID: 'abcde1234',
            accessToken: clientInfo.extensionConfiguration?.accessToken ?? 'invalid',
            serverEndpoint: clientInfo.extensionConfiguration?.serverEndpoint ?? dotcom,
            customHeaders: {},
        })
    })

    it.skip('lists recipes correctly', async () => {
        const recipes = await client.request('recipes/list', null)
        assert.equal(9, recipes.length, JSON.stringify(recipes))
    })

    const filePath = path.join(workspaceRootPath, 'src', 'main.ts')
    const uri = Uri.file(filePath)
    it('accepts textDocument/didOpen notifications', async () => {
        const content = await fspromises.readFile(filePath, 'utf8')
        client.notify('textDocument/didOpen', {
            uri: uri.toString(),
            content,
            selection: { start: { line: 4, character: 0 }, end: { line: 4, character: 0 } },
        })
    })

    it.skip('returns non-empty autocomplete', async () => {
        const completions = await client.request('autocomplete/execute', {
            uri: uri.toString(),
            position: { line: 4, character: 0 },
            triggerKind: 'Invoke',
        })
        const texts = completions.items.map(item => item.insertText)
        expect(completions.items.length).toBeGreaterThan(0)
        expect(texts).toMatchInlineSnapshot(`
          [
            "   return a + b;",
          ]
        `)
        client.notify('autocomplete/completionAccepted', { completionID: completions.items[0].id })
    }, 10_000)

    const messages: ExtensionMessage[] = []
    client.registerNotification('webview/postMessage', ({ message }) => {
        messages.push(message)
    })

    it.skip('allows us to execute recipes properly', async () => {
        const id = await client.request('chat/new', null)
        const messageID = uuid.v4()
        await client.request('webview/receiveMessage', {
            id,
            message: {
                command: 'submit',
                text: 'Hello',
                submitType: 'user',
                addEnhancedContext: true,
                contextFiles: [],
                messageID,
            },
        })
        await new Promise<void>(resolve => setTimeout(resolve, 1000))
        console.log('foo')
        expect(messages.map(message => message.type)).toMatchInlineSnapshot(`
          {
            "contextFiles": [],
            "preciseContext": [],
            "speaker": "assistant",
            "text": " Hello! I'm Cody, an AI assistant created by Sourcegraph to help with programming tasks. I don't have any context about the codebase or your questions yet, but I'm ready to assist you based on the information you provide. I won't make any assumptions or provide hypothetical examples without proper context. Please feel free to share code snippets or details about what you're working on, and I'll do my best to provide helpful answers!",
          }
        `)
    }, 20_000)

    // TODO Fix test - fails intermittently on macOS on Github Actions
    // e.g. https://github.com/sourcegraph/cody/actions/runs/7191096335/job/19585263054#step:9:1723
    it.skip('allows us to cancel chat', async () => {
        setTimeout(() => client.notify('$/cancelRequest', { id: client.id - 1 }), 300)
        await client.request('recipes/execute', { id: 'chat-question', humanChatInput: 'How do I implement sum?' })
    }, 600)

    afterAll(async () => {
        await client.shutdownAndExit()
        // Long timeout because to allow Polly.js to persist HTTP recordings
    }, 20_000)
})
