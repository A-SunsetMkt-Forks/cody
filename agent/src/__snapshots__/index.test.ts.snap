// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Agent > Chat > chat/submitMessage (long message) 1`] = `
"I'll create a simple Hello World function in Java for you.

\`\`\`java:src/HelloWorld.java
public class HelloWorld {
    public static void main(String[] args) {
        sayHello();
    }

    public static void sayHello() {
        System.out.println("Hello, World!");
    }
}
\`\`\`

To compile and run this Java program, you can use these commands:

\`\`\`bash
javac src/HelloWorld.java
\`\`\`

\`\`\`bash
java -cp src HelloWorld
\`\`\`

This creates a clean and simple Hello World program with a dedicated function to print the greeting. The program uses a separate \`sayHello()\` method which demonstrates good practice for function organization. Feel free to modify the message or add more functionality!"
`;

exports[`Agent > Chat > chat/submitMessage (short message) 1`] = `
{
  "model": "anthropic::2024-10-22::claude-3-5-sonnet-latest",
  "speaker": "assistant",
  "text": "Hi there! Great to meet you! I'm Cody, and I'm excited to help you with any coding or technical questions you may have. What would you like to work on?",
}
`;

exports[`Agent > Chat > chat/submitMessage (with mock context) 1`] = `
"\`\`\`typescript:src/dog.ts
export class Dog implements Animal {
    name: string;
    isMammal = true;

    constructor(name: string) {
        this.name = name;
    }

    makeAnimalSound(): string {
        return "Woof!";
    }
}
\`\`\`"
`;

exports[`Agent > Commands > commands/explain 1`] = `
"Sure. Here's an explanation for the provided code:

Code: Animal Interface in TypeScript

Purpose:
The purpose of this code is to define the structure for objects that represent animals in a program. This helps ensure that all objects that represent animals in the program have the same properties and methods, making it easier to work with them.

Inputs:
The code doesn't take any inputs. Instead, it defines the properties and methods that objects representing animals in the program should have.

Outputs:
The code doesn't produce any output. Instead, it defines a structure for objects that represent animals in the program.

Logic:
The code defines an interface, which is like a blueprint for an object. The interface has three properties: "name", "makeAnimalSound", and "isMammal". The "name" property is a string that represents the name of the animal. The "makeAnimalSound" property is a method that returns a string representing the sound that the animal makes. The "isMammal" property is a boolean that indicates whether the animal is a mammal.

Explanation:

* The first line, \`export interface Animal {\`, makes the \`Animal\` interface available for use in other files.
* The next three lines, \`name: string\`, \`makeAnimalSound(): string\`, and \`isMammal: boolean\`, define the properties that objects representing animals in the program should have. The \`name\` property is a string that represents the name of the animal, while the \`makeAnimalSound\` property is a method that returns a string representing the sound that the animal makes. The \`isMammal\` property is a boolean that indicates whether the animal is a mammal.
* The last line, \`}\`, closes the interface definition.

How it achieves its purpose:
The code defines an interface, which is a template for objects. Objects that implement this interface will have the properties and methods specified in the interface, ensuring consistency and predictability in working with these objects.

Important Logic Flows or Data Transformations:
There are no complex logic flows or data transformations in this code. The code simply defines a structure for objects that represent animals in a program.

Overall, this code defines the structure for objects that represent animals in a program, making it easier to work with these objects and ensuring that they have consistent properties and methods."
`;

exports[`Agent > Commands > commands/smell 1`] = `
"Based on the provided code:

\`\`\`typescript
export interface Animal {
    name: string
    makeAnimalSound(): string
    isMammal: boolean
}
\`\`\`

There are a few suggestions I would like to propose to improve the code's readability, maintainability, and alignment with best practices:

1. **Type Annotations**: Add type annotations to the interface properties. This can improve the code's readability and self-documentation.
    \`\`\`typescript
    export interface Animal {
        name: string;
        makeAnimalSound(): string;
        isMammal: boolean;
    }
    \`\`\`
2. **Method Documentation**: Provide a brief documentation comment for the interface and each method to explain their purpose. This can help developers understand the code more easily and improve maintainability.
    \`\`\`typescript
    /**
     * An interface representing an animal with name, a method to make its sound,
     * and whether it's a mammal.
     */
    export interface Animal {
        /** The name of the animal. */
        name: string;
        /** A method that generates the animal's sound. */
        makeAnimalSound(): string;
        /** A boolean value indicating if the animal is a mammal. */
        isMammal: boolean;
    }
    \`\`\`
3. **Exported Interfaces**: Align the interface properties for readability. This can help ensure the code consistently follows best practices.
    \`\`\`typescript
    export interface Animal {
        name: string;
        isMammal: boolean;
        makeAnimalSound(): string;
    }
    \`\`\`
4. **Access Modifiers**: Consider adding explicit access modifiers. Although this is a TypeScript/JavaScript convention rather than a strict requirement, adding access modifiers such as "public" or "private" can help define the interface's intended behavior more clearly.
    \`\`\`typescript
    export interface Animal {
        /** The name of the animal. */
        public name: string;
        /** A boolean value indicating if the animal is a mammal. */
        public isMammal: boolean;
        /** A method that generates the animal's sound. */
        public makeAnimalSound(): string;
    }
    \`\`\`
5. **Code Style**: Follow a consistent code style, such as using single quotes instead of double quotes for string literals. Although this is a subjective point, consistency in style can improve readability and maintainability.
    \`\`\`typescript
    export interface Animal {
        name: string;
        isMammal: boolean;
        makeAnimalSound(): string;
    }
    \`\`\`

Summarizing the analysis, I identified several opportunities for improvement in the given code, mainly related to readability, maintainability, and best practices. Considering these suggestions can help ensure a more robust, efficient, and consistently designed codebase."
`;
